sqldelight-idea-plugin/src/main/kotlin/app/cash/sqldelight/intellij/run/SelectConnectionTypeAction.kt

```diff
@@ -0,0 +1,119 @@
+package app.cash.sqldelight.intellij.run
+
+import app.cash.sqldelight.intellij.util.dialectPreset
+import app.cash.sqldelight.intellij.util.isSqlite
+import com.intellij.ide.util.PropertiesComponent
+import com.intellij.ide.util.propComponentProperty
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.fileChooser.FileTypeDescriptor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.DialogWrapper
+import com.intellij.openapi.ui.ValidationInfo
+import com.intellij.ui.RecentsManager
+import com.intellij.ui.TextFieldWithHistoryWithBrowseButton
+import com.intellij.ui.layout.ValidationInfoBuilder
+import com.intellij.ui.layout.buttonGroup
+import com.intellij.ui.layout.panel
+import com.intellij.ui.layout.selected
+import javax.swing.JComponent
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+internal const val RECENT_DB_PATH = "app.cash.sqldelight.recentPath"
+internal const val DB_CONNECTION_TYPE = "app.cash.sqldelight.connectionType"
+
+internal class SelectConnectionTypeAction : AnAction() {
+
+  override fun actionPerformed(e: AnActionEvent) {
+    val project = e.project
+    if (project != null) {
+      SelectConnectionTypeDialog(project).show()
+    }
+  }
+
+  override fun update(e: AnActionEvent) {
+    val project = e.project ?: return
+    e.presentation.isEnabledAndVisible = project.dialectPreset.isSqlite
+  }
+}
+
+internal enum class ConnectionType(val type: Int) {
+  NONE(0),
+  FILE(1)
+}
+
+private class ConnectionTypeProperty(project: Project) : ReadWriteProperty<Any, ConnectionType> {
+  private val propertiesComponent = PropertiesComponent.getInstance(project)
+
+  override fun getValue(thisRef: Any, property: KProperty<*>): ConnectionType {
+    val type = propertiesComponent.getInt(DB_CONNECTION_TYPE, ConnectionType.NONE.type)
+    return ConnectionType.values().firstOrNull { it.type == type } ?: ConnectionType.NONE
+  }
+
+  override fun setValue(thisRef: Any, property: KProperty<*>, value: ConnectionType) {
+    propertiesComponent.setValue(
+      DB_CONNECTION_TYPE,
+      value.type,
+      ConnectionType.NONE.type
+    )
+  }
+}
+
+private fun connectionType(project: Project): ReadWriteProperty<Any, ConnectionType> =
+  ConnectionTypeProperty(project)
+
+internal class ConnectionOptions(val project: Project) {
+  var filePath: String by propComponentProperty(project, "")
+  var connectionType: ConnectionType by connectionType(project)
+}
+
+internal class SelectConnectionTypeDialog(
```

Comment:
okay and then this is the thing where it would be sweet if we could somehow find localhost ports that have a valid db running to connect to. If thats even possible we can probably just skim IntelliJ source code to find something that already does it, but again not for this PR